conjure-java-undertow-runtime/src/main/java/com/palantir/conjure/java/undertow/runtime/ConjureError.java

```diff
@@ -0,0 +1,100 @@
+import com.palantir.conjure.java.api.errors.CheckedServiceException;
+import com.palantir.conjure.java.api.errors.ErrorType;
+import com.palantir.conjure.java.api.errors.RemoteException;
+import com.palantir.conjure.java.api.errors.SerializableError;
+import com.palantir.conjure.java.api.errors.ServiceException;
+import com.palantir.logsafe.Arg;
+import com.palantir.logsafe.Preconditions;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A representation of a Conjure defined error that will be serialized and sent over the wire from a server to a client.
+ * <p>Previously {@link SerializableError} was used for this purpose. The main difference is that
+ * {@link SerializableError#parameters()} is a {@code Map<String, String>} while {@link ConjureError#parameters} is a
+ * {@code Map<String, Object>}. When constructing a {@link SerializableError} from a {@link ServiceException}, the error
+ * parameters were converted to strings using {@link Objects#toString} which made it difficult for the client to recover
+ * the structured type. Storing the parameter value as an object allows servers to send Jackson-serialized objects which
+ * the client can convert to their original type.
+ *
+ * @param errorCode A fixed code word identifying the type of error. For errors generated from {@link ServiceException}
+ *                  and {@link CheckedServiceException}, this corresponds to the {@link ErrorType#code} and is part of
+ *                  the service's API surface.
+ * @param errorName A fixed name identifying the error. For errors generated from {@link ServiceException} and
+ *                  {@link CheckedServiceException}, this corresponding to the {@link ErrorType#name} and is part of the
+ *                  service's API surface.
+ * @param errorInstanceId errorInstanceId A unique identifier for (this instance of) this error.
+ * @param parameters params A list of parameters providing additional context about the error.
+ */
+record ConjureError(
+        @JsonProperty("errorCode") String errorCode,
+        @JsonProperty("errorName") String errorName,
+        @JsonProperty("errorInstanceId") String errorInstanceId,
+        @JsonProperty("parameters") @JsonInclude(content = Include.NON_ABSENT) Map<String, Object> parameters) {
+
+    ConjureError {
+        Preconditions.checkNotNull(errorCode, "errorCode cannot be null");
+        Preconditions.checkNotNull(errorName, "errorName cannot be null");
+        Preconditions.checkNotNull(errorInstanceId, "errorInstanceId cannot be null");
+        Preconditions.checkNotNull(parameters, "parameters cannot be null");
+    }
+
+    static ConjureError fromCheckedServiceException(CheckedServiceException exception) {
+        Map<String, Object> parameters = new HashMap<>();
+        for (Arg<?> arg : exception.getArgs()) {
+            if (arg.getValue() == null) {
+                continue;
+            }
```

Comment:
I was thinking of the case where service developers can add `Optional` fields to their error definitions without causing a break so long as they send `Optional.empty` until clients roll onto a new version of their API.

But we get even better back-compat safety by having the client deserializer ignore unknown properties. I think using a deserializer which ignores unknown properties sounds reasonable. Adding a whole new error that an endpoint can throw is certainly an API break and should fail loudly. While the fields in an error are part of the API surface, I'm not sure we need to have clients fail if they encounter a field in an error they haven't seen.

Reply:
Doesn't `@JsonInclude(content = Include.NON_ABSENT)` take care of this for us already? The same applies to Optional.empty, OptionalInt.empty, etc.

> But we get even better back-compat safety by having the client deserializer ignore unknown properties. I think using a deserializer which ignores unknown properties sounds reasonable

Our clients use [ObjectMappers. newClientJsonMapper()]https://github.com/palantir/conjure-java-runtime/blob/349e914a8620a5141c78da2f2ea6a48698f71d25/conjure-java-jackson-serialization/src/main/java/com/palantir/conjure/java/serialization/ObjectMappers.java#L52-L56 which sets `.disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)` for precisely this reason -- this is true of all response types :-)

Reply:
> Doesn't @JsonInclude(content = Include.NON_ABSENT) take care of this for us already?

It didn't. I think it has to do with the fact that the value is `Object`. The `NON_ABSENT` flag worked when I used an `Optional` as the value. I wrote a custom contentFilter.